addons/mail/tools/background_task.py

```diff
@@ -0,0 +1,149 @@
+import logging
+
+import threading
+from queue import Queue, Full
+
+from odoo import api
+from odoo.modules.registry import Registry
+from odoo.service.server import CommonServer
+
+_logger = logging.getLogger(__name__)
+
+
+class Task:
+    def __init__(self, func, env, *args, **kwargs):
+        self.func = func
+        self.env = env
+        self.args = args
+        self.kwargs = kwargs
+
+class BackgroudTaskQueue(Queue[Task]):
+    ...
+
+class BackgroundTaskWorker(threading.Thread):
+    """
+    Background task to handle non-critical operations
+    """
+
+    def __init__(self, queue: BackgroudTaskQueue, id: int):
+        super().__init__(daemon=True, name=f'BackgroundTaskWorker-{id}')
+        _logger.info(f"Starting BackgroundTaskWorker-{id}")
+        self._queue = queue
+        self.max_error_count = 10
+        self.task_error_count = 0
+        self.task_count = 0
+        self.error_count = 0
+
+
+    def loop(self):
+        while True:
+            task = self._queue.get()
+            try:
+                db_registry = Registry(task.env["dbname"])
+                with db_registry.cursor() as cr:
+                    env = api.Environment(cr, task.env['uid'], task.env['context'])
+                    task.func(task.env["self"].with_env(env), *task.args, **task.kwargs)
+                self.task_count += 1
+            except Exception as e:
+                _logger.error(f"Error in background task: {e}")
+                self.error_count += 1
+            finally:
+                self._queue.task_done()
+            self.task_error_count = 0
```

Comment:
Does it serve a purpose to reset `task_error_count` after each iteration, while it's only incremented outside of the loop? I think it could be reset after the call to `self.loop()` in `run()` so that all references to `task_error_count` are next to each other.

Also it seems that it will only reach `max_error_count` if errors are consecutive (but maybe it's intended).